### PR TITLE
RFC about using Symfony UX Turbo

### PR DESCRIPTION
**Pull request by jrmgx**
Hello, and thank you for your work.

In my project I already use Symfony UX Turbo, so when I choose to use Dynamic Forms (and implement the JavaScript part) I saw an opportunity to have both working together.

In this PR I updated the README, but it's more to have a place to show you the idea and implementation. 

This may even lead to a better integration between this package and Turbo Frame.

Let me know.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -153,9 +153,90 @@ This library doesn't handle this for you, but here are the 2 main options:
 This is the easiest method: by rendering your form inside a live component,
 it will automatically re-render when the form changes.
 
-### B) Write custom JavaScript
+### B) Use [Symfony UX Turbo](https://symfony.com/bundles/ux-turbo/current/index.html#decomposing-complex-pages-with-turbo-frames)
+
+By using turbo frames you can have a dynamic form running quickly without any JavaScript.
```
Comment (bocharsky-bw) [inline README.md:158]:
```suggestion
By using Turbo Frames you can have a dynamic form running quickly without any JavaScript.
```

File: README.md [hunk 2]
```diff
@@ -153,9 +153,90 @@ This library doesn't handle this for you, but here are the 2 main options:
...
+<turbo-frame id="rating-form">
+{{ form_start(form) }}
+    {{ form_row(form.rating) }}
+
+    {% if form.badRatingNotes is defined %}
+        {{ form_row(form.badRatingNotes) }}
+    {% endif %}
+    {# We also add a real submit button #}
+    {{ form_row(form.submit) }}
+{{ form_end(form) }}
```
Comment (bocharsky-bw) [inline README.md:172]: Let's indent everything in `turbo-frame` tag with a tab please :)

Comment (bocharsky-bw) [reply README.md:172]: Btw, I don't see any value of showing the expanded form rendering here. wha about to simplify the whole example to just `{{ form(form) }}`? It would already mean that you just need to render your form in that spot. If you have custom rending with `form_row()` etc - that's up to you. Because right now it seems like we're focusing on rendering something important between that `form_start`/`form_end` to make it work, but that's not true if I understand it correctly

File: README.md (hunk 1 above)
Comment (bocharsky-bw) [reply README.md:158]: Hm, should we also mention that you have to use Turbo on your website? Right now it sounds like I just need to write the code below and it should magically work, but if people do not have Turbo activated on that page - that will not work, right?

**Review by bocharsky-bw — COMMENTED** [review]
I like this idea in general, though have some comments

File: README.md (hunk 2 above)
Comment (jrmgx) [reply README.md:172]: Right, the only thing important is to have a proper `submit` button, but as it is specified in the `FormType`, there is no need to have it fully expanded here.

I'll simplify it!

File: README.md (hunk 1 above)
Comment (jrmgx) [reply README.md:158]: Yes indeed, and more than that, we need to explain that only Turbo Frame is needed, not Turbo Drive (many of my colleague do not know the difference). I'll rephrase it to make it clear.

**Comment by jrmgx** [comment]
I have fixed your comments, please let me know if it's good for you.

File: README.md [hunk 3]
```diff
@@ -155,21 +155,19 @@ it will automatically re-render when the form changes.
...
-{{ form_start(form) }}
-    {{ form_row(form.rating) }}
-
-    {% if form.badRatingNotes is defined %}
-        {{ form_row(form.badRatingNotes) }}
-    {% endif %}
-    {# We also add a real submit button #}
-    {{ form_row(form.submit) }}
-{{ form_end(form) }}
+{{ form(form) }}
```
Comment (bocharsky-bw) [inline README.md:170]: I like this form simplification example. The only point we want to mention here - users need to render the form. How - that's up to them, and Symfony docs about Forms help with it.

But let's indent it.
```suggestion
    {{ form(form) }}
```

File: README.md [hunk 4]
```diff
@@ -153,9 +153,88 @@ This library doesn't handle this for you, but here are the 2 main options:
...
+            'Okay' => 3,
+            'Bad' => 2,
+            'Terrible' => 1
+        ],
+        // This will allow the form to auto-submit on value change
+        'attr' => ['onchange' => 'this.form.submit()'],
+    ]);
+    // This will allow to differenciate between a user submition and an auto-submit
+    $builder->add('submit', SubmitType::class, [
+        'attr' => ['value' => 'submit'], // Needed for turbo
```
Comment (bocharsky-bw) [inline README.md:197]:
```suggestion
        'attr' => ['value' => 'submit'], // Needed for Turbo
```

File: README.md [hunk 5]
```diff
@@ -153,9 +153,88 @@ This library doesn't handle this for you, but here are the 2 main options:
...
+
+```twig
+<turbo-frame id="rating-form">
+{{ form(form) }}
+</turbo-frame>
+```
+
+From here you need two small changes:
+
+First, in your form type:
```
Comment (bocharsky-bw) [inline README.md:176]: let's describe with words what exactly you need to do in your form type... it's not pretty clear from the code IMO.

Also, I think we can use GitHub's `diff` code block to show what lines of code were added. Wdyt?

File: README.md [hunk 6]
```diff
@@ -153,9 +153,88 @@ This library doesn't handle this for you, but here are the 2 main options:
...
+        'attr' => ['onchange' => 'this.form.submit()'],
+    ]);
+    // This will allow to differenciate between a user submition and an auto-submit
+    $builder->add('submit', SubmitType::class, [
+        'attr' => ['value' => 'submit'], // Needed for turbo
+    ]);
+
+```
+
+Second, in your controller:
```
Comment (bocharsky-bw) [inline README.md:202]: Same here, I would like to mention what exactly devs should do in their controllers and why. And `diff` code block instead of PHP may help to notice the change easier. Thoughts?

Comment (jeromegxj) [reply README.md:202]: Agreed for more detail textual description of the changes, but for the `diff` format, if used, we lose the `php` highlighting, and it's kind of hard to read then (or maybe we can use both? diff+php).

Comment (bocharsky-bw) [reply README.md:202]: Hm, not sure `diff+php` will work, though never tried it. Yeah, with diff will lose PHP highlighting, but if we're talking about a few new lines that should be added in a big code block - I'm fine with it. It would help to focus on specific lines

File: README.md [hunk 7]
```diff
@@ -153,9 +153,114 @@ This library doesn't handle this for you, but here are the 2 main options:
...
++           'action' => $this->generateUrl('feedback'),
++       ]);
+        $feedbackForm->handleRequest($request);
+        if ($feedbackForm->isSubmitted() && $feedbackForm->isValid()) {
+
++           /** @var SubmitButton $submitButton */
++           $submitButton = $feedbackForm->get('submit');
++           if (!$submitButton->isClicked()) {
++               return $this->render('feedback.html.twig', ['feedbackForm' => $feedbackForm]);
++           }
```
Comment (bocharsky-bw) [inline README.md:248]: Isn't this should be below "Your code here"? I mean, we still want to execute all that custom code in below, but just return a different reponse, right?

File: README.md [hunk 8]
```diff
@@ -153,9 +153,114 @@ This library doesn't handle this for you, but here are the 2 main options:
...
+
+            // Your code here
+            // ...
+
+            return $this->redirectToRoute('home');
```
Comment (bocharsky-bw) [inline README.md:253]: And if so, maybe better to do the opposite.... if `$submitButton->isClicked()` is clicked - do this `return $this->redirectToRoute('home');`.

Otherwise, the execution will continue and call that `return $this->render('feedback.html.twig', ['feedbackForm' => $feedbackForm]);` below.

Wdyt?

File: README.md (hunk 7 above)
Comment (jrmgx) [reply README.md:248]: In this scenario, this code is here to handle the `'onchange' => 'this.form.submit()'` (auto-submit) part of the form. So it should not trigger the business code, it just re-renders the form with updated state.

**Review by bocharsky-bw — APPROVED** [review]
OK, I think this is good to go!

Thank you for working on it!
